python/ray/serve/_private/deployment_state.py

```diff
@@ -2004,39 +2004,32 @@ def _get_all_node_ids(self):
         # Test mock purpose
         return get_all_node_ids(self._gcs_client)
 
-    def _deploy_driver(self) -> bool:
+    def _deploy_driver(self) -> List[ReplicaSchedulingRequest]:
         """Deploy the driver deployment to each node."""
-        all_nodes = self._get_all_node_ids()
-        deployed_nodes = set()
-        for replica in self._replicas.get(
-            [
-                ReplicaState.STARTING,
-                ReplicaState.RUNNING,
-                ReplicaState.RECOVERING,
-                ReplicaState.UPDATING,
-                ReplicaState.STOPPING,
-            ]
-        ):
-            if replica.actor_node_id:
-                deployed_nodes.add(replica.actor_node_id)
-        replica_changed = False
-        for node_id, _ in all_nodes:
-            if node_id in deployed_nodes:
-                continue
+        upscale = []
+        num_existing_replicas = self._replicas.count()
+        if num_existing_replicas >= self._target_state.num_replicas:
+            num_running_replicas = self._replicas.count(states=[ReplicaState.RUNNING])
+            if num_running_replicas >= self._target_state.num_replicas:
+                for replica in self._replicas.pop(states=[ReplicaState.STARTING]):
+                    self._stop_replica(replica)
```

Comment:
shouldn't this logic live in the scheduler rather than here as part of the downscale request?

Reply:
It's not the normal downscaling (driver deployment doesn't downscale based on qps). It's about cancelling extra replicas.

I added some comments to make it clear:
```
# Cancel starting replicas when driver deployment state creates
# more replicas than alive nodes.
# For example, get_all_node_ids returns 4 nodes when
# the driver deployment state decides the target number of replicas
# but later on when the deployment scheduler schedules these 4 replicas,
# there are only 3 alive nodes (1 node dies in between).
# In this case, 1 replica will be in the PENDING_ALLOCATION and we
# cancel it here.
```